### DISCUSSION TITLE
HDF5 Output

### DISCUSSION THREAD

**Discussion by richardvanderburgh** [AMReX-Codes/amrex#3326] [Q&A]
Hello AMReX community!

AMReX is new to me and I am trying to produce .h5 output files. I have successfully built AMReX with MPI and HDF5 enabled (using CMake on Windows), but I am not sure if I am correctly use the H5 plotting functions (WriteSingleLevelPlotfileHDF5 andWriteMultiLevelPlotfileHDF5).

I could not find any examples in the tutorial for outputting .h5 files, so I replaced the WriteSingleLevelPlotfile call with WriteSingleLevelPlotfile in the Heat_Equation_EX1_C main.cpp file.

After this change I am getting these linker errors:

2>Auto build dll exports
2>   Creating library .................amrex-tutorials/ExampleCodes/build/_deps/amrex-build/Src/Debug/amrex_3d.lib and object ...........amrex-tutorials/ExampleCodes/build/_deps/amrex-build/Src/Debug/amrex_3d.exp
2>AMReX_PlotFileUtilHDF5.obj : error LNK2001: unresolved external symbol H5T_C_S1_g
2>AMReX_PlotFileUtilHDF5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_INT_g
2>AMReX_PlotFileUtilHDF5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_LLONG_g
2>AMReX_PlotFileUtilHDF5.obj : error LNK2001: unresolved external symbol H5T_NATIVE_DOUBLE_g
2>AMReX_PlotFileUtilHDF5.obj : error LNK2001: unresolved external symbol H5P_CLS_FILE_ACCESS_ID_g
2>AMReX_PlotFileUtilHDF5.obj : error LNK2001: unresolved external symbol H5P_CLS_DATASET_CREATE_ID_g
2>AMReX_PlotFileUtilHDF5.obj : error LNK2001: unresolved external symbol H5P_CLS_DATASET_XFER_ID_g

Any and all help is greatly appreciated!

Much Thanks!

**Comment by atmyers** [comment]
I haven't seen this before, but I see this discussion re: similar linking errors on Windows: https://github.com/conda-forge/hdf5-feedstock/issues/58

Could you try adding `-DH5_BUILT_AS_DYNAMIC_LIB` to the C++ flags, as suggested in that answer?

**Reply by richardvanderburgh** [reply]
Thank you for the help! I built the HDF5 library from the source provided on the HDF5 Group's website. Should I put this flag in the library's CMakeLists.txt file? Also, I tried using conda-forge to install hdf5, but I could not get manage to get a version that has MPI. Thanks again.

**Reply by atmyers** [reply]
I think it's saying you want to add this define when building AMReX against the HDF5 library you installed. Maybe something like the following patch to `Tools/CMake/AMReXSetDefines.cmake`:

```
diff --git a/Tools/CMake/AMReXSetDefines.cmake b/Tools/CMake/AMReXSetDefines.cmake
index c15f7bb23..2ccd40d94 100644
--- a/Tools/CMake/AMReXSetDefines.cmake
+++ b/Tools/CMake/AMReXSetDefines.cmake
@@ -173,6 +173,9 @@ add_amrex_define(AMREX_USE_HDF5 NO_LEGACY IF AMReX_HDF5)
 add_amrex_define(AMREX_USE_HDF5_ASYNC NO_LEGACY IF AMReX_HDF5_ASYNC)
 add_amrex_define(AMREX_USE_HDF5_ZFP NO_LEGACY IF AMReX_HDF5_ZFP)
 
+if(WIN32)
+    add_amrex_define(H5_BUILT_AS_DYNAMIC_LIB NO_LEGACY IF AMReX_HDF5)
+endif()
 
 #
 # SUNDIALS
```

I don't have access to a Windows machine so I can't test this easily, you may have to play around with that a bit to get it to work.
